File: backend/app/schemas/audit.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


class AuditLogActor(BaseModel):
    id: Optional[str]
    email: Optional[str]
    first_name: Optional[str]
    last_name: Optional[str]


class AuditLogEntry(BaseModel):
    id: str
    action_type: str
    target_type: str
    target_id: Optional[str]
    metadata: Optional[dict[str, Any]]
    created_at: datetime
    actor: Optional[AuditLogActor]


class AuditLogListResponse(BaseModel):
    items: list[AuditLogEntry]
    page: int
    page_size: int
    total: int

File: frontend/src/lib/api/audit.ts
import { apiClient, toApiError } from "@/lib/api/client";

export type AuditLogActor = {
  id: string | null;
  email: string | null;
  first_name: string | null;
  last_name: string | null;
};

export type AuditLogEntry = {
  id: string;
  action_type: string;
  target_type: string;
  target_id: string | null;
  metadata: Record<string, unknown> | null;
  created_at: string;
  actor: AuditLogActor | null;
};

export type AuditLogListResponse = {
  items: AuditLogEntry[];
  page: number;
  page_size: number;
  total: number;
};

export type AuditLogFilters = {
  page?: number;
  pageSize?: number;
  actionType?: string;
  startAt?: string;
  endAt?: string;
};

export async function fetchAuditLogs(filters: AuditLogFilters = {}): Promise<AuditLogListResponse> {
  try {
    const { data } = await apiClient.get<AuditLogListResponse>("/admin/audit/logs", {
      params: {
        page: filters.page,
        page_size: filters.pageSize,
        action_type: filters.actionType || undefined,
        start_at: filters.startAt || undefined,
        end_at: filters.endAt || undefined
      }
    });
    return data;
  } catch (error) {
    throw toApiError(error);
  }
}
